### ISSUE TITLE
First SideEffect doesn't get called in React 15

### ISSUE THREAD

**Issue by rickyvetter** [reasonml/reason-react#130]
The React component rendered on the `simple` page of this repo exhibits different behavior when installed with NPM (v4 and v5) and Yarn. With NPM the first click shows:

```
Click count: 0
Switch state: on
```

With Yarn it shows (correctly):

```
Click count: 1
Switch state: on
```

The click count is updated in a side effect. There are no obvious differences between toplevel dependencies on NPM/Yarn.

https://github.com/rickyvetter/reason-react-example/tree/npmyarn

**Comment by jordwalke** [comment]
I cannot possibly imagine how the package manager could result in a difference here. Are there any errors in the console?

**Comment by rickyvetter** [comment]
No errors or warnings at build time or run time.

**Comment by rickyvetter** [comment]
I've confirmed the issue here is a different version of React being resolved in NPM versus Yarn. [`yarn why`](https://gist.github.com/rickyvetter/fc40f8a178ac193f77db769c425398de#file-gistfile1-txt-L468) shows a React 16 dependency nested under ReasonReact which is separate from the toplevel React 15. NPM resolves both to React 15. It appears that SideEffects has a bug only present in React 15. Not sure if this behavior existed before 0b1d7633fb72af724c32d73f0def0f499e71da48. cc @cristianoc

Steps I took to confirm:
1. Removing nested dependency manually from yarn install and verifying that I can reproduce the problem. I think we should fiddle with our dependencies a little to see if we can avoid this difference between Yarn/NPM (cc @chenglou)
2. Bumping to 16.1.1 and installing with NPM. Verifying that I can no longer repro the issue.
